sdks/go/pkg/beam/core/metrics/metrics.go

```diff
@@ -82,15 +82,19 @@ func (ctx *beamCtx) Value(key interface{}) interface{} {
 	switch key {
 	case bundleKey:
 		if ctx.bundleID == "" {
-			if id := ctx.Value(key); id != nil {
+			if id := ctx.Context.Value(key); id != nil {
```

Comment:
This is where the infinite recursion was happening, using ctx.Context.Value is the fix!
Value is a method on the context.Context interface which is ordinarily promoted to beamCtx via embedding. But we're overriding it on beamCtx. In other words, the Value being called was the one we're looking at, instead of the embedded Context's Value. In short, if the value didn't exist, it would recursively call itself forever.
As you've noted, calling ctx.Context.Value makes it explicit we want the embedded types implementation, rather than the beamCtx.Value. There's no benefit in this case to call beamCtx.Value since we already know this level doesn't have the bundleID (or ptransformID).

Reply:
Ah, that makes sense now. The piece I missed was that the current function was beamCtx.Value.